pennylane/transforms/optimization/single_qubit_fusion.py

```diff
@@ -0,0 +1,120 @@
+# Copyright 2018-2021 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Transforms for optimizing quantum circuits."""
+
+from pennylane import numpy as np
+from pennylane import apply
+from pennylane.transforms import qfunc_transform
+from pennylane.ops.qubit import Rot
+from pennylane.math import allclose, cast_like, stack, zeros
+
+from .optimization_utils import find_next_gate, fuse_rot_angles
+
+
+@qfunc_transform
+def single_qubit_fusion(tape, tol=1e-8):
+    """Quantum function transform to fuse together groups of single-qubit
+    operations into the general single-qubit unitary form (:class:`~.Rot`).
```

Comment:
I like the idea of having `gate_set`, it would be similar to what we did with `merge_rotations`, where you can specify the set of gates that should be merged. I can see this having utility in benchmarking / error mitigation; I also know of hardware that has implemented things like the identity gate with two explicit pulses that cancel each other (`RX(pi/2), RX(-pi/2)` or something like this), and so you wouldn't want to fuse those gates.  

I think we need to decide though whether we want things specified by inclusion or exclusion because I feel it'd be confusing if different transforms (with a similar effect) have different behaviour of their arguments. For this case, since there are so many single-qubit gates, you'd want to specify which gates to exclude rather than include, whereas for rotation merging there aren't as many gates so either one is fine, but at least for me it feels more natural there that you'd pass which gates to include in the transform.

Reply:
> I think we need to decide though whether we want things specified by inclusion or exclusion because I feel it'd be confusing if different transforms (with a similar effect) have different behaviour of their arguments.

I think we can side-step this issue if we are explicit in naming. E.g., if it makes more sense for one to be exclusion and the other to be inclusion, we could have `exclude_gates={...}` and `include_gates={...}`.